server.py

```diff
@@ -190,12 +190,29 @@ def announce():
 		server["total_clients"] = server["clients"]
 	server["pop_v"] = server["total_clients"] / server["updates"]
 
+	old_err = errorTracker.get(getErrorPK(server))
+
 	finishRequestAsync(server)
 
+	if old_err:
+		return ("Request has been filed, "
+			"but the previous request encountered the following error:\n" +
+			old_err, 409)
 	return "Request has been filed.", 202
 
 # Utilities
 
+# returns a primary key suitable for saving and replaying an error unique to a
+# server that was announced.
+def getErrorPK(server):
+	# We need to include the client IP in here, since some failures
+	# only happen depending on it.
+	return "%s/%s/%d" % (server["ip"], server["address"], server["port"])
```

Comment:
You shouldn't have to do string interpolation, you should directly be able to use a tuple as a key.

Reply:
true